Refactor flask factory ( closes Ohtu-FaceTed/FaceTed-Search#88 )

alafuzof opened the pull request:
Large scale reorganization of Flask code. The main change is shift to factory pattern for Flask app creation, so that we can easily have a different config (e.g. database) for testing and production use. Also cleaned up some of the dependency hell between files.

alafuzof committed:
Moved views into separate folder and refactored them to use blueprints.

alafuzof committed:
Moved database initialization to separate module (cf. views)
MODIFIED
> src/__init__.py
> src/models/__init__.py

alafuzof committed:
Refactored question selection to be independent of session data.
MODIFIED
> src/question_selection.py
> src/views/answer.py
> src/views/previous.py
> src/views/question.py
> tests/test_question_selection.py

alafuzof committed:
Refactored main __init__ to present a factory function instead of a global app
object.
MODIFIED
> app.py
> src/__init__.py
> src/models/__init.py
> tests/test_api.py

alafuzof committed:
Refactored configuration code into separate objects. Testing now uses different
configuration (incl. database) to production.
ADDED
> src/config.py
MODIFIED
> .gitignore
> app.py
> config.py
> src/__init__.py
> tests/tests_api.py

alafuzof committed:
Linting for all of the previous work on the branch.
MODIFIED
> src/__init__.py
> src/config.py
> src/models/__init__.py
> src/question_selection.py
> src/views/__init__.py
> src/views/index.py
> src/views/question.py
> tests/test_question_selection.py

alafuzof commented:
Note: as far as I know, there is no way around the unused import Codacy is complaining about. This is because the route decorator only registers the functions as routes upon import of the module.

alafuzof committed:
Removed unused function in question selection
MODIFIED
> src/question_selection.py

EssKayz reviewed (commented):
Checked through, seems fine. Will download branch later today and test it locally a bit!

vaisanem committed:
suggested changes to configuration
facilitated the usage of production secret key
thought extending Config object was unnecessary and moved both to /config.py

vaisanem reviewed (commented):
Made a few changes, but approved on my part
